### DISCUSSION TITLE
Fan-out durable functions consume excessive amount of CPU

### DISCUSSION THREAD

**Discussion by dxynnez** [Azure/azure-functions-durable-extension#2568] [Q&A]
Hello team,

We have a function app running on dedicated app service plan (I1V2 with 3 instances). We recently added a durable function which would fan-out to ~100 sub orchestration functions, and eventually results in ~ 200 activity functions (1 sub-orch to 2 activitys). We tried to concurrently 'enqueue' 800 such durable functions and observed dramatic performance degradation and excessive amount of CPU usage.

Each activity function seems to execute relatively fast (~ 30 - 155ms), but the sub orchestrations seem to take 400s to complete on average, which eventually slow down the completion of the top-level orchestrations (took 950s to complete). The CPU usage on each worker spikes to > 80% when the durable functions are running.

We took a profiler trace on the worker and most of the CPU clocks seem to be spent on durable storage related modules (> 50%), while the CPU spent on our function module was relatively small (~ 7%).

![image](https://github.com/Azure/azure-functions-durable-extension/assets/10041558/dcbc182a-82b7-47a5-9042-76f44410d60f)![image](https://github.com/Azure/azure-functions-durable-extension/assets/10041558/079b1404-7767-450f-aa11-b3b16781eb64)

Given our scale (800 concurrent durables * 100 sub-orchestraction * 2 activity function each) and 3 I1V2 workers, is this expected (perf and resource utilization wise)? Anything we can do to improve it (besides scaling up & out the ASP)?

Some other information that might help:

We are on .NET 6 Azure function V4 in-process mode with Microsoft.Azure.WebJobs.Extensions.DurableTask 2.11.0

We are using the default host.json, but we did try to tweak it a little bit by:
 1) reducing the concurrency level of the activity function & orchestration function (as i1v2 only has 2 cores)
 2) increase the control queue buffer
 3) enable extended session & increase extended session idle timeout

But the default still seems to be the most performant setting.

**Comment by cgillum** [comment]
400s (6-7 minutes) for a sub-orchestration seems really excessive. The only reason for performance this bad is that something is crashing. I would definitely look into the memory usage of your app. I also recommend going through our [troubleshooting guide](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-troubleshooting-guide), if you haven't already. Note the part about using [Azure Function app diagnostics](https://learn.microsoft.com/en-us/azure/azure-functions/durable/function-app-diagnostics) to help diagnose common problems.

What size payloads are you passing between activities and sub-orchestrations? Sending large payloads can have a dramatic impact on performance.

Lastly, you should not be running any production apps on Functions v3, since that version is out of support. Please upgrade your runtime to Functions V4 ASAP.

**Reply by dxynnez** [reply]
@cgillum thanks for looking into this!

Memory usage of the app is quite normal (~500M), and the Azure function app diagnostics doesn't show any obvious problems except the CPU usage.

We did go through some of the troubleshooting guide and tried things like lowering the concurrency / enabling extended session (30s idle timeout) / increasing the ControlQueueBufferThreshold (to 512) but none of them really change anything (well lowering the concurrency did lower the CPU usage and 'unfreeze' the worker to certain degree). We currently don't use app insights, but I can turn that on and run the test again to collect some telemetry - do you think that would help?

We checked the storage table & blob and there wasn't anything, so I assume the payloads were all handled by the storage queue.

For function v3, that was my mistake, we are indeed running on V4.

**Reply by cgillum** [reply]
Turning on application insights might help you dive more deeply into what's going on, but may not be super useful for diagnosing CPU issues.

> We checked the storage table & blob and there wasn't anything, so I assume the payloads were all handled by the storage queue.

When you said, "there wasn't anything", did you mean that literally (like no tables or data)? There should always be data there, even if your payloads are small.

How many actions do your orchestrations perform? By chance could it be that your histories are growing really large? For example, do you have any orchestrations that might be scheduling thousands of tasks over its lifetime?

**Reply by dxynnez** [reply]
@cgillum 

There were some data, but nothing related to large payload (we did reviewed one of your recommendation and figured that might be the cause of the poor performance and hence we changed our implementation to fire 800 durable function concurrently instead of using a single durable function to fan-out to 80000 activity functions).

Now each top level orchestration will fire 200 sub-orchestrations and each sub-orchestration only execute 2 activities - but the result indicates that the top level orchestration, even with around 200 activities (the sub-orch) to perform, still execute relatively fast (~950s to complete which is still linear to the sub-orch which took ~400s). But the sub-orch, although only triggers 2 activities, takes ~400s to complete on average, while all its activities only took ~30-155ms to complete.

The payload of the sub-orch and its activities should be less than 100B.

**Reply by cgillum** [reply]
OK, so that helps narrow down the problem, which is something about the activities and not about orchestration processing generally. I could see two possibilities:

1. The work-item queue (which schedules activity calls) is really backed up or
2. The activities periodically crash, adding ~300s (5 minute) delays.

For (1), you could check the Azure Storage queue length to see if this might be the case. For (2), you should be able to detect this by looking at the troubleshooting tools, or by looking at the Durable Task Framework logs for Azure Storage and checking to see whether the value for `DequeueCount` is ever 2 or more.

**Reply by dxynnez** [reply]
Thanks @cgillum !
For 1), I checked the [storage Queue Message Count metric](https://microsoft.visualstudio.com/CMD/_git/W365C-Redist-HostingCore?path=/src/Core/Diagnostics/ConfigureLoggingFilterOptions.cs&version=GBdevelop&line=22&lineEnd=23&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) and it's showing ~18K message during the particular hour (1hr is the minimum granularity that the metric supports) - is that a lot? If so, does it mean we need more powerful workers / more workers to handle the load?

For 2), that shouldn't be the case as our telemetry is showing very low to none failure rate for the activities, the function app's Function Executions and Errors detector is indicating the same thing.

We also observed another weird behavior that, even for executing a single durable function (which fans out to 4 sub orchestrations), it took around 25s to complete. Our telemetry is indicating that the sub orchestrations were sort of executed in sequence - they were executed on different workers but never concurrently with around 5-10s difference (assuming no clock skew here). Is this normal?

**Reply by cgillum** [reply]
~18K messages is a lot. You'll need to consider a) increasing activity function concurrency (particularly if resource utilization is low), b) scaling out to a larger number of VMs, or c) using more powerful VMs (if resource utilization is high and you're not able to scale-out further).

> Our telemetry is indicating that the sub orchestrations were sort of executed in sequence...Is this normal?

I'd need to look at your code to confirm whether you correctly scheduled the sub-orchestrations to run in parallel, but assuming you did, I'd expect that they actually run in parallel, regardless of whether they're on the same VM or different VMs. You may want to take a look at increasing your maximum orchestration concurrency as well if you think they delay might be load-related, just like it is for your activity functions.

**Reply by dxynnez** [reply]
Thanks @cgillum !

` You'll need to consider a) increasing activity function concurrency (particularly if resource utilization is low), b) scaling out to a larger number of VMs, or c) using more powerful VMs (if resource utilization is high and you're not able to scale-out further).`
I guess this basically implies that the situation is completely normal as our CPU is constantly hitting >90%？Do we have any formula to calculate how many durable functions / activity functions (dry run) that the durable framework can handle, with an average VM? Does 3000 executions / min (this is what's showing on the Function Execution Count metric, after we scheduled the 800 durables which fan-out to 80000 activities in total) sound like a lot, assuming all the activities are basically dry-run and would exit right after it's triggered?

For fan-out, we just add the task returned by the IDurableOrchestrationContext and doing a Task.WhenAll at this end, something like this:
```
  foreach (var input in inputs)
  {
      Task task= context.CallSubOrchestratorAsync(nameof(#functionName#, input);
      allTasks.Add(task);
  }
  await Task.WhenAll(allTasks);
```
This shouldn't be load-related as we really only triggered a single durable function when testing. The default orchestration concurrency (cores * 10) should be enough to handle all the sub-orchestrations (4). Also, per my understanding the orchestration concurrency is per machine, but in our cases the sub-orchestrations are executed by different VMs, but the VMs seem to pickup those fan-out sub-orchestrations in sequence.

**Reply by dxynnez** [reply]
poking @cgillum . Any more pointer? Appreciate it!

**Comment by dxynnez** [comment]
Hello @cgillum , is there anything else we should look into?

**Comment by xsurfer** [comment]
is there any update on this thread? I’m interested since facing a similar huge CPU usage

**Comment by quido** [comment]
My understanding is that hitting the CPU hard via CPU-intensive loops and such causes some sort of storage resource to crash and retry after a delay. https://github.com/Azure/azure-functions-durable-extension/issues/1687

In my durable function, I read a ton of data via API, and this works great. Then when I hit a loop to build 10k large objects in memory with this data, the activity or something supporting it crashes dozens of times before finally getting lucky and succeeding usually 40-60 minutes later. I've been trying to throttle the CPU usage with Thread.Sleep(), but it only seems to make it worse in the cloud.
